core/src/transfer/data.rs

```diff
@@ -39,6 +43,126 @@ pub enum TransactionData {
     /// Additional data added to a transaction, that is not a deployment or a
     /// call.
     Memo(Vec<u8>),
+    /// Data for blob storage together with contract call.    
+    Blob(Vec<BlobData>),
+}
+
+impl TransactionData {
+    /// Return input bytes to hash the `TransactionData`.
+    ///
+    /// Note: The result of this function is *only* meant to be used as an input
+    /// for hashing and *cannot* be used to deserialize the payload again.
+    #[must_use]
+    pub fn signature_message(&self) -> Vec<u8> {
+        let mut bytes = vec![];
+
+        #[allow(clippy::match_same_arms)]
+        match &self {
+            TransactionData::Deploy(d) => {
+                bytes.extend(&d.bytecode.to_hash_input_bytes());
+                bytes.extend(&d.owner);
+                if let Some(init_args) = &d.init_args {
+                    bytes.extend(init_args);
+                }
+            }
+            TransactionData::Call(c) => {
+                bytes.extend(c.contract.as_bytes());
+                bytes.extend(c.fn_name.as_bytes());
+                bytes.extend(&c.fn_args);
+            }
+            TransactionData::Memo(m) => {
+                bytes.extend(m);
+            }
+            TransactionData::Blob(blobs) => {
+                // Hash the blobs and extend the signature_message with the
+                // result, using the same serialization format used by
+                // TransactionData::Memo, so that the legacy transfer contract
+                // can still verify it after being converted with `blob_to_memo`
+                let mut hasher = Sha256::new();
+                for blob in blobs {
+                    hasher.update(blob.to_hash_input_bytes());
```

Comment:
Also, are we sure it is sufficient to sign the versioned hash, without the proof and the data?
Is this by eip scheme?

Reply:
The reason for the signature to only include the hash is to still be able to verify the signature once the sidecar has been deleted (this might be needed, for instance, when a node is syncing the blockchain).
If we sign the blob sidecar and then delete the sidecar, we will not be able to verify the signature anymore.

Instead, what should be done is to verify the validity of commitment and proof when validating a block (and before entering the mempool)